File: Check If a String Contains All Binary Codes of Size K/Check If a String Contains All Binary Codes of Size K.java
// Solution for Check If a String Contains All Binary Codes of Size K in JAVA

class Solution {
    public boolean hasAllCodes(String s, int k) {

        int totalCodesNeeded = 1 << k;
        Set<String> seen = new HashSet<>();

        for (int i = k; i <= s.length(); i++) {

            String sub = s.substring(i - k, i);

            if (!seen.contains(sub)) {
                seen.add(sub);
                totalCodesNeeded--;
            }

            if (totalCodesNeeded == 0)
                return true;
        }

        return false;
    }
}

File: Check If a String Contains All Binary Codes of Size K/Check If a String Contains All Binary Codes of Size K.py
# Solution for Check If a String Contains All Binary Codes of Size K in PY

class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:

        total_codes_needed = 1 << k
        seen = set()

        for i in range(k, len(s) + 1):

            sub = s[i - k:i]

            if sub not in seen:
                seen.add(sub)
                total_codes_needed -= 1

            if total_codes_needed == 0:
                return True

        return False
